executor/src/witgen/jit/compiler.rs

```diff
@@ -0,0 +1,539 @@
+    match effect {
+        Effect::Assignment(cell, _) => iter::once(cell),
+        Effect::RangeConstraint(..) => unreachable!(),
+        Effect::Assertion(..) => iter::empty(),
+        Effect::MachineCall(_, arguments) => arguments.iter().flat_map(|e| match e {
+            MachineCallArgument::Unknown(e) => Some(e.single_unknown_variable().unwrap()),
+            MachineCallArgument::Known(_) => None,
+        }),
+    }
+}
+
+fn format_effect<T: FieldElement>(effect: &Effect<T, Cell>) -> String {
+    match effect {
+        Effect::Assignment(var, e) => {
+            format!(
+                "    let {} = {};",
+                cell_to_var_name(var),
+                format_expression(e)
+            )
+        }
+        Effect::RangeConstraint(..) => {
+            unreachable!("Final code should not contain pure range constraints.")
+        }
+        Effect::Assertion(Assertion {
+            lhs,
+            rhs,
+            expected_equal,
+        }) => format!(
+            "    assert!({} {} {});",
+            format_expression(lhs),
+            if *expected_equal { "==" } else { "!=" },
+            format_expression(rhs)
+        ),
+        Effect::MachineCall(id, arguments) => todo!(),
+    }
+}
+
+fn format_expression<T: FieldElement>(e: &SymbolicExpression<T, Cell>) -> String {
+    match e {
+        SymbolicExpression::Concrete(v) => format!("FieldElement::from({v})"),
+        SymbolicExpression::Symbol(cell, _) => cell_to_var_name(cell),
+        SymbolicExpression::BinaryOperation(left, op, right, _) => {
+            let left = format_expression(left);
+            let right = format_expression(right);
+            match op {
+                BinaryOperator::Add => format!("({left} + {right})"),
+                BinaryOperator::Sub => format!("({left} - {right})"),
+                BinaryOperator::Mul => format!("({left} * {right})"),
+                BinaryOperator::Div => format!("({left} / {right})"),
+                BinaryOperator::IntegerDiv => format!("integer_div({left}, {right})"),
+                BinaryOperator::BitAnd => format!("({left} & {right})"),
+                BinaryOperator::BitOr => format!("({left} | {right})"),
+            }
+        }
+        SymbolicExpression::UnaryOperation(op, inner, _) => {
+            let inner = format_expression(inner);
+            match op {
+                UnaryOperator::Neg => format!("-{inner}"),
+            }
+        }
+    }
+}
+
+fn cell_to_var_name(
+    Cell {
+        column_name,
+        row_offset,
+        ..
+    }: &Cell,
+) -> String {
+    // TODO this might still not be unique.
+    let column_name = column_name
+        .rfind("::")
+        .map_or(column_name.as_str(), |i| &column_name[i + 2..])
+        .replace("[", "_")
+        .replace("]", "_");
+    if *row_offset < 0 {
+        format!("{column_name}_u{}", -row_offset)
+    } else {
+        format!("{column_name}_d{row_offset}")
+    }
```

Comment:
Why u and d?

Reply:
`up` and `down`. Please suggest a better name :)

Reply:
Don't have a great idea. Maybe I'd just prepend `n` for "negative", and nothing for the offsets >= 0 (the typical case).